allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -36,101 +38,23 @@ def index_with(self, vocab: Vocabulary):
 class AllennlpLazyDataset(IterableDataset):
     def __init__(
         self,
-        deserialize: Callable[[str], Instance] = None,
-        serialize: Callable[[Instance], str] = None,
-        vocab: Vocabulary = None,
-    ) -> None:
-        super().__init__()
-        self.deserialize = deserialize
-        self.serialize = serialize
-        self.vocab = vocab
-
-    def index_with(self, vocab: Vocabulary):
-        self.vocab = vocab
-
-
-class _LazyInstances(AllennlpLazyDataset):
-    """
-    An `Iterable` that just wraps a thunk for generating instances and calls it for
-    each call to `__iter__`.
-    """
-
-    def __init__(
-        self,
-        instance_generator: Callable[[str], Iterable[Instance]],
+        instance_iterator_factory: Callable[[str], Iterable[Instance]],
```

Comment:
This could use a docstring now it is more complicated, highlighting that you use this like:

`AllennlpLazyDataset(DatasetReader(lazy=True).read)`

Reply:
Can we just pass a dataset reader class instead? Or at least make that an option?

I feel strongly that the word "factory" should appear nowhere in our public API. Passing functions is a dodgy concept for many users. Passing members is worse. Passing members that return lazy generators is 🤯.  Sometimes these things are unavoidable, but I don't think this is one of those times.